kamu-core/src/domain/query_data_accessor.rs

```diff
@@ -0,0 +1,29 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use datafusion::prelude::SessionContext;
+use opendatafabric::{DataSlice, DatasetAlias};
+
+use super::InternalError;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+pub trait QueryDataAccessor: Send + Sync {
```

Comment:
I quite like how DataFusion separates the concerns:
- All objects are referenced by URLs
- To read/write/list objects you need an `ObjectStore`
- There can be many `ObjectStore` instances for different locations
- Location to `ObjectStore` mapping is done based on URL prefixes

Perhaps we can follow a similar separation in our code?

For example:
- Add `Dataset::get_layout(&self) -> DatasetLayout` method
- `DatasetLayout` will allow resolving data/checkpoint/block hashes to full object URLs
  - Note that this abstraction allows datasets in one `DatasetRegistry` still potentially to be in different S3 buckets, different regions, spread between S3/GCS/Azure, or even have checkpoints and data be in different locations
- Separately we introduce some sort of `ObjectStoreRegistry` that `QueryService` can use to get all available `ObjectStore`s and register them in `DataFusion` before executing queries.

Reply:
I'm imagining something like:
```rust
let dataset = dataset_repo.get_dataset(dataset_ref)?;
let data_slice_url = dataset.get_layout().get_data_slice_url(&data_slice);
let object_store = datafusoin_ctx.runtime_env().object_store(data_slice_url)?;
let object_metadata = object_store.head(data_slice_url).await?;
```

Reply:
Also have a look at [`datafusion::ObjectStoreRegistry` docs](https://docs.rs/datafusion/25.0.0/datafusion/datasource/object_store/trait.ObjectStoreRegistry.html).

> Systems relying on ad-hoc discovery, without corresponding DDL, can create `ObjectStore` lazily by providing a custom implementation of `ObjectStoreRegistry`

This might be useful to connect DataFusion to whatever new type we will have that provides object store implementations.